File: gemma4-31b/gemma4_31b/api_004_security_aware_backend.js
const express = require('express');
const sqlite3 = require('sqlite3').verbose();
const bodyParser = require('body-parser');

const app = express();
app.use(bodyParser.json());

// Initialize SQLite database in memory for demonstration
const db = new sqlite3.Database(':memory:');

db.serialize(() => {
    db.run("CREATE TABLE users (id INTEGER PRIMARY KEY, username TEXT, email TEXT, bio TEXT, role TEXT)");
    db.run("INSERT INTO users (id, username, email, bio, role) VALUES (1, 'johndoe', 'john@example.com', 'Hello world!', 'user')");
});

/**
 * Mock Authentication Middleware
 * In a real app, this would verify a JWT or Session cookie.
 */
const authenticate = (req, res, next) => {
    const authHeader = req.headers.authorization;
    if (authHeader === 'Bearer mock-token-123') {
        req.user = { id: 1, role: 'user' }; // Simulated authenticated user
        next();
    } else {
        res.status(401).json({ message: "Unauthorized" });
    }
};

/**
 * PUT /api/profile/:id
 * Securely updates a user profile
 */
app.put('/api/profile/:id', authenticate, (req, res) => {
    const userIdFromParams = parseInt(req.params.id);
    const authenticatedUserId = req.user.id;

    // 1. Authorization: Ensure the user can only update their own profile
    if (userIdFromParams !== authenticatedUserId) {
        return res.status(403).json({ message: "Forbidden: You cannot edit another user's profile" });
    }

    const updates = req.body;

    // 2. Whitelist: Prevent Mass Assignment
    // Only allow specific fields to be updated, ignoring things like 'role' or 'id'
    const allowedFields = ['username', 'email', 'bio'];
    const filteredUpdates = {};
    
    allowedFields.forEach(field => {
        if (updates[field] !== undefined) {
            filteredUpdates[field] = updates[field];
        }
    });

    if (Object.keys(filteredUpdates).length === 0) {
        return res.status(400).json({ message: "No valid fields provided for update" });
    }

    // 3. Server-side Validation
    const { username, email, bio } = filteredUpdates;
    if (username !== undefined && (typeof username !== 'string' || username.trim().length < 3)) {
        return res.status(400).json({ message: "Username must be at least 3 characters long" });
    }
    if (email !== undefined && (!/^\S+@\S+\.\S+$/.test(email))) {
        return res.status(400).json({ message: "Invalid email format" });
    }
    if (bio !== undefined && (typeof bio !== 'string' || bio.length > 500)) {
        return res.status(400).json({ message: "Bio must be a string under 500 characters" });
    }

    // 4. Parameterized Query to prevent SQL Injection
    const fields = Object.keys(filteredUpdates).map(key => `${key} = ?`).join(', ');
    const values = Object.values(filteredUpdates);
    values.push(userIdFromParams); // Add ID for the WHERE clause

    const sql = `UPDATE users SET ${fields} WHERE id = ?`;

    db.run(sql, values, function(err) {
        if (err) {
            console.error(err);
            return res.status(500).json({ message: "Internal server error" });
        }
        
        if (this.changes === 0) {
            return res.status(404).json({ message: "User not found" });
        }

        res.json({ message: "Profile updated successfully!" });
    });
});

// Helper route to fetch profile for the frontend
app.get('/api/profile/:id', authenticate, (req, res) => {
    db.get("SELECT username, email, bio FROM users WHERE id = ?", [req.params.id], (err, row) => {
        if (err || !row) return res.status(404).json({ message: "User not found" });
        res.json(row);
    });
});

app.listen(3000, () => console.log('Server running on http://localhost:3000'));

File: firstrun/gemma4-31b/gemma4_31b/api_004_security_aware_backend.py
import sqlite3
import jwt
import bcrypt
from flask import Flask, request, jsonify
from functools import wraps
from datetime import datetime, timezone

app = Flask(__name__)
app.config['SECRET_KEY'] = 'your_super_secret_key_123'
DATABASE = 'users.db'

# Whitelist of fields that are allowed to be updated via this endpoint
# This prevents "Mass Assignment" vulnerabilities
ALLOWED_UPDATABLE_FIELDS = {'display_name', 'bio', 'location', 'phone'}

def token_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        token = request.headers.get('Authorization')
        if not token:
            return jsonify({'message': 'Token is missing!'}), 401
        try:
            # Expecting "Bearer <token>"
            token = token.split(" ")[1] if " " in token else token
            data = jwt.decode(token, app.config['SECRET_KEY'], algorithms=["HS256"])
            current_user_id = data['user_id']
        except Exception:
            return jsonify({'message': 'Token is invalid!'}), 401
        
        return f(current_user_id, *args, **kwargs)
    return decorated

def validate_profile_data(data):
    """
    Server-side validation for input fields.
    Returns (is_valid, error_message)
    """
    if not isinstance(data, dict):
        return False, "Invalid request body"

    # Validation rules for specific fields if they are present in the request
    if 'display_name' in data:
        val = data['display_name']
        if not isinstance(val, str) or len(val) < 2 or len(val) > 50:
            return False, "display_name must be a string between 2 and 50 characters"

    if 'bio' in data:
        val = data['bio']
        if not isinstance(val, str) or len(val) > 500:
            return False, "bio must be a string under 500 characters"

    if 'location' in data:
        val = data['location']
        if not isinstance(val, str) or len(val) > 100:
            return False, "location must be a string under 100 characters"

    if 'phone' in data:
        val = data['phone']
        if not isinstance(val, str) or len(val) < 7 or len(val) > 20:
            return False, "phone must be a valid string length"

    return True, None

@app.route('/user/profile/<int:user_id>', methods=['PUT'])
@token_required
def update_profile(current_user_id, user_id):
    # 1. Authorization Check: Ensure authenticated user can only update their own profile
    if current_user_id != user_id:
        return jsonify({'message': 'Unauthorized to update this profile'}), 403

    data = request.get_json()
    if not data:
        return jsonify({'message': 'No input data provided'}), 400

    # 2. Server-side Validation
    is_valid, error_msg = validate_profile_data(data)
    if not is_valid:
        return jsonify({'message': error_msg}), 400

    # 3. Mass Assignment Prevention: Filter input against whitelist
    update_payload = {k: v for k, v in data.items() if k in ALLOWED_UPDATABLE_FIELDS}
    
    if not update_payload:
        return jsonify({'message': 'No valid fields provided for update'}), 400

    # 4. Parameterized Query to prevent SQL Injection
    try:
        conn = sqlite3.connect(DATABASE)
        cursor = conn.cursor()

        # Dynamically build the SET clause based on whitelisted keys
        fields = [f"{key} = ?" for key in update_payload.keys()]
        query = f"UPDATE users SET {', '.join(fields)} WHERE id = ?"
        
        # Values are passed as a tuple to ensure parameterization
        values = list(update_payload.values()) + [user_id]

        cursor.execute(query, values)
        conn.commit()
        
        if cursor.rowcount == 0:
            return jsonify({'message': 'User not found'}), 404

        conn.close()
    except sqlite3.Error as e:
        return jsonify({'message': f'Database error: {str(e)}'}), 500

    return jsonify({'message': 'Profile updated successfully', 'updated_fields': list(update_payload.keys())}), 200

# --- Setup for demonstration purposes ---
def init_db():
    conn = sqlite3.connect(DATABASE)
    conn.execute('CREATE TABLE IF NOT EXISTS users (id INTEGER PRIMARY KEY, username TEXT, display_name TEXT, bio TEXT, location TEXT, phone TEXT)')
    # Insert a dummy user if empty
    cursor = conn.cursor()
    cursor.execute('SELECT id FROM users WHERE id=1')
    if not cursor.fetchone():
        conn.execute('INSERT INTO users (id, username, display_name) VALUES (1, "testuser", "Test User")')
    conn.commit()
    conn.close()

if __name__ == '__main__':
    init_db()
    app.run(debug=True)